src/ucs/memory/memtype_cache.c

```diff
@@ -201,22 +201,36 @@ UCS_PROFILE_FUNC(ucs_status_t, ucs_memtype_cache_lookup,
                  ucs_memtype_cache_t *memtype_cache, const void *address,
                  size_t size, ucs_memory_type_t *mem_type_p)
 {
-    const ucs_pgt_addr_t start = (uintptr_t)address;
+    ucs_pgt_addr_t start       = (uintptr_t)address;
+    const ucs_pgt_addr_t end   = start + size;
+    ucs_memory_type_t mem_type = UCS_MEMORY_TYPE_LAST;
     ucs_memtype_cache_region_t *region;
     ucs_pgt_region_t *pgt_region;
     ucs_status_t status;
 
     pthread_rwlock_rdlock(&memtype_cache->lock);
 
-    pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup, &memtype_cache->pgtable,
-                                  start);
-    if ((pgt_region == NULL) || (pgt_region->end < (start + size))) {
-        status = UCS_ERR_NO_ELEM;
-        goto out_unlock;
-    }
+    do {
+        pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup,
+                                      &memtype_cache->pgtable,
+                                      start);
+        if (pgt_region == NULL) {
+            status = UCS_ERR_NO_ELEM;
+            goto out_unlock;
+        }
+
+        region = ucs_derived_of(pgt_region, ucs_memtype_cache_region_t);
+        ucs_assert((mem_type == UCS_MEMORY_TYPE_LAST) ||
+                   (region->mem_type == mem_type));
```

Comment:
i think we should return some error which would fail UCP (not assertion) if memory type is not same in all regions
and in general, maybe merge regions with same memtype, and if we have only partial intersection return UCS_MEMORY_TYPE_LAST

Reply:
> i think we should return some error which would fail UCP (not assertion) if memory type is not same in all regions

ok, but if we will do merging - we don't need this loop

> and in general, maybe merge regions with same memtype,

let's do it

> and if we have only partial intersection return UCS_MEMORY_TYPE_LAST

i.e. if `ucs_pgtable_lookup` returns valid region, but `pgt_region->end < (start + size)` is false, we will return `UCS OK` and set `UCS_MEMORY_TYPE_LAST`, right?

Reply:
> > i think we should return some error which would fail UCP (not assertion) if memory type is not same in all regions
> 
> ok, but if we will do merging - we don't need this loop

I think we need it in this PR, instead of failing on assertion

> 
> > and in general, maybe merge regions with same memtype,
> 
> let's do it
> 
> > and if we have only partial intersection return UCS_MEMORY_TYPE_LAST
> 
> i.e. if `ucs_pgtable_lookup` returns valid region, but `pgt_region->end < (start + size)` is false, we will return `UCS OK` and set `UCS_MEMORY_TYPE_LAST`, right?

Yes, because we can't let the user think the memory type is "host" in such case.